drivers/flash/flash_stm32h7x.c

```diff
@@ -0,0 +1,531 @@
+			regs->KEYR1 = FLASH_KEY1;
+			regs->KEYR1 = FLASH_KEY2;
+		}
+	}
+#ifdef DUAL_BANK
+	/* Bank 2 */
+	if (enable) {
+		regs->CR2 |= FLASH_CR_LOCK;
+	} else {
+		if (regs->CR2 & FLASH_CR_LOCK) {
+			regs->KEYR2 = FLASH_KEY1;
+			regs->KEYR2 = FLASH_KEY2;
+		}
+	}
+#endif
+
+	if (enable) {
+		LOG_DBG("Enable write protection");
+	} else {
+		LOG_DBG("Disable write protection");
+	}
+
+	flash_stm32_sem_give(dev);
+
+	return rc;
+}
+
+static void flash_stm32h7_flush_caches(const struct device *dev,
+				       off_t offset, size_t len)
+{
+	ARG_UNUSED(dev);
+	SCB_InvalidateDCache_by_Addr((uint32_t *)(CONFIG_FLASH_BASE_ADDRESS
+						  + offset), len);
+}
+
+
+static int flash_stm32h7_erase(const struct device *dev, off_t offset,
+			       size_t len)
+{
+	int rc;
+	off_t flush_offset = get_sector(offset) * FLASH_SECTOR_SIZE;
+	size_t flush_len = (((get_sector(offset + len) + 1) * FLASH_SECTOR_SIZE)
+			    - flush_offset);
+
+	if (!flash_stm32_valid_range(dev, offset, len, false)) {
+		LOG_ERR("Erase range invalid. Offset: %ld, len: %zu",
+			(long) offset, len);
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	flash_stm32_sem_take(dev);
+
+	LOG_DBG("Erase offset: %ld, len: %zu", (long) offset, len);
+
+	rc = flash_stm32_block_erase_loop(dev, offset, len);
+
+	/* Flush cache on all sectors affected by the erase */
+	flash_stm32h7_flush_caches(dev, flush_offset, flush_len);
+
+	flash_stm32_sem_give(dev);
+
+	return rc;
+}
+
+
+static int flash_stm32h7_write(const struct device *dev, off_t offset,
+			       const void *data, size_t len)
+{
+	int rc;
+
+	if (!flash_stm32_valid_range(dev, offset, len, true)) {
+		LOG_ERR("Write range invalid. Offset: %ld, len: %zu",
+			(long) offset, len);
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	flash_stm32_sem_take(dev);
+
+	LOG_DBG("Write offset: %ld, len: %zu", (long) offset, len);
+
+	rc = flash_stm32_write_range(dev, offset, data, len);
+
+	flash_stm32_sem_give(dev);
+
+	return rc;
+}
+
+static int flash_stm32h7_read(const struct device *dev, off_t offset,
+			      void *data,
+			      size_t len)
+{
+	if (!flash_stm32_valid_range(dev, offset, len, false)) {
+		LOG_ERR("Read range invalid. Offset: %ld, len: %zu",
+			(long) offset, len);
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	LOG_DBG("Read offset: %ld, len: %zu", (long) offset, len);
+
+	memcpy(data, (uint8_t *) CONFIG_FLASH_BASE_ADDRESS + offset, len);
+
+	return 0;
+}
+
+
+static const struct flash_parameters flash_stm32h7_parameters = {
+	.write_block_size = FLASH_STM32_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+static const struct flash_parameters *
+flash_stm32h7_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_stm32h7_parameters;
+}
+
+static const struct flash_pages_layout stm32h7_flash_layout[] = {
+	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
+#ifdef DUAL_BANK
+	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
```

Comment:
I think you will face the same issue I have on STM32G4 #30231
There may be some discontinuity between bank1 and bank2 on some chip, when dual bank, with smaller flash .
Ex: stm32h747zi has 2M flash but stm32h747IG has only 1M flash, in this case real number of sector will be smaller, and by the way you cannot rely on FLASH_SECTOR_SIZE which is defined to 8 for all soc STM32H747xx.

Reply:
I've had a look at files under `modules/hal/stm32/stm32cube/stm32h7xx/soc`, and all SOCs defining `DUAL_BANK` have a 1MB `FLASH_BANK_SIZE` (based on the comments).
So as far as I understand it the h7 family would not be impacted by this fragmented issue. What do you think ?

Reply:
Unfortunately, it seems thoses files are erroneous. As I said, stm32h747IG has only 1M flash (512 HB per bank).
I submitted an internal ticket to change FLASH_SIZE definition.
And so h7 family is impacted by potential discontinuity.
As FLASH_SIZE (as well as FLASH_BANK_SIZE) is not reliable for now, you can use `LL_GetFlashSize()`

Reply:
@nicovince, after discussion with @erwango,
we propose you to not implement discontinuity management in this PR,
and I will do it in a separate PR.

Reply:
Ok, my bad.

I'll look into it then. How can I test my code on this particular chip ?